gson/src/main/java/com/google/gson/internal/bind/TypeAdapterRuntimeTypeWrapper.java

```diff
@@ -51,7 +52,7 @@ public void write(JsonWriter out, T value) throws IOException {
 
     TypeAdapter chosen = delegate;
     Type runtimeType = getRuntimeTypeIfMoreSpecific(type, value);
-    if (runtimeType != type) {
+    if (runtimeType != type && !Primitives.areEquivalentTypes(runtimeType, type)) {

```

Comment:
There is no other way that I can think of. The problem is that ReflectiveTypeAdapterFactory (line 109) field.get(value) returns Long even though the field is of type long. This results in the runtime type being incorrectly characterized for all primitive types.  Hence the TypeAdapterRuntimeTypeWrapper needs to be smarter when dealing with primitive wrapper types.